---
title: "Testing"
output:
  pdf_document:
    latex_engine: lualatex
  html_document:
    df_print: paged
  html_notebook: default
---

```{r include = FALSE}
knitr::opts_chunk$set( cache = TRUE )
```


Alright, let's see if the revised script makes sense.
```{r, cache = FALSE}
source( "~/Script repository/R scripting/Matrix_Fns.R" )
```


```{r genes}
prox <- read.delim( 
  file = "C:/Users/kayle/Box/Vertinolab/McKayla Ford/Projects/R-loops_Pausing/Sequence_based_analysis/Classing_by_Skew/beds/prox_tss_order.bed",
  header = FALSE
  )
dist <- read.delim(
  file = "C:/Users/kayle/Box/Vertinolab/McKayla Ford/Projects/R-loops_Pausing/Sequence_based_analysis/Classing_by_Skew/beds/dist_cgi_order.bed",
  header = FALSE
  )
```

```{r labelling}
gene_info <- rbind( prox, dist )
colnames( gene_info ) <- c( "uniq_id", "chrom", "cpg_s", "cpg_e",
                            "cpg_num", "gene_s", "gene_e",
                            "strand", "name", "ex2in", "gene_l",
                            "cpg_l", "TSS_3prime", "prime5_TSS",
                            "TSS_ex", "TSS_skew", "CGI3_skew",
                            "TES_skew", "CGI5_skew", "x2in_skew",
                            "simple_skew_class")
```


```{r making beds}
#my debug uses chr 19 and 22 so lets get a sample of that.
gene_info_chr <- gene_info[ ( gene_info$chrom == "chr19" | gene_info$chrom == "chr22"  ), ]
gene_info_p <- gene_info_chr[ gene_info_chr$strand == "+", ]
gene_info_m <- gene_info_chr[ gene_info_chr$strand == "-", ]

gi_p_rand <- randsamp( gene_info_p, s = 20 )
gi_m_rand <- randsamp( gene_info_m, s = 20 )

TSS_CGI_E_p <- gi_p_rand[ , c("chrom", "gene_s", "cpg_e", "uniq_id", "TSS_skew", "strand") ]

TSS_CGI_E_m <- gi_m_rand[ , c("chrom", "cpg_s", "gene_e", "uniq_id", "TSS_skew", "strand") ]

TSS_CGI_E_p
TSS_CGI_E_m
```


Now to try them out. I know my scripts are currently broken so I'll start by
trialling pieces. We need to do a bam and a bigwig for each type for proseq and something non-stranded, like Zach's cut and tag.

```{r}
Dir <- "C:/Users/kayle/Box/Vertinolab/McKayla Ford/Data/"
bam1 <- paste0( Dir, "/CutnTag_n_chip/Vertino lab data/VER00046 2021.11.17 Cut and Tag/bams/hg38_dedup/ZS1_19_MCF7_EV_H3K27me3.dedup.bam" )
bam2 <- paste0( Dir, "/NascentSeq/PublicData/GSE93229_Danko_2018_MCF7_PROseq/bam/bam_nodedup/MCF7_H9.bam" )
bw1 <-  paste0( Dir, "/CutnTag_n_chip/Vertino lab data/VER00046 2021.11.17 Cut and Tag/bigwigs/prededup_bamcoveragefix/ZS1_19_MCF7_EV_H3K27me3.prededup.rpkm.bw" )
bw2p <- paste0( Dir, "/NascentSeq/PublicData/GSE93229_Danko_2018_MCF7_PROseq/bw/pauseloc/MCF7_H9_fwd.bw" )
bw2m <- paste0( Dir, "/NascentSeq/PublicData/GSE93229_Danko_2018_MCF7_PROseq/bw/pauseloc/MCF7_H9_rev.bw" )
```


Let's try bv bamvars.

```{r}
bv1 <- bam_vars( bam1, pairedEnd = TRUE, rnorm = TRUE )
bv2 <- bam_vars( bam2, pairedEnd = FALSE, rnorm = FALSE )
```

```{r}
bv1[[1]]
bv1[[2]]
bv1[[3]]
bv2[[1]]
bv2[[2]]
bv2[[3]]
```
All look good.
```{r}
chrs <- list( "chr19", "chr22" )
bed_subp <- TSS_CGI_E_p[ TSS_CGI_E_p[[1]] == "chr19", ]
bed_subm <- TSS_CGI_E_m[ TSS_CGI_E_m[[1]] == "chr19", ]
```
Let's try the dif mats.
```{r}
a <- 150
b <- 50
bs <- 20
n <- 10 

hist_center_p <- non_scaled_mat( bed_sub = bed_subp, a = a, b = b, bs = bs, method = "peak_centered" )

hist_sanch_p <- non_scaled_mat( bed_sub = bed_subp, a = a, b = b, bs = bs, method = "single_anch" )
strandvec_sanch_p <- rep( bed_subp[[6]], times = ( a + b ) / bs )
      
mat_list <- bi_anch_mat( bed_sub = bed_subp, n = n )
hist_scaled_p <- mat_list[[1]]
ends_matp <- mat_list[[2]]
strandvec_scaled_p <- rep( bed_subp[[6]], times = n )  

hist_center_m <- non_scaled_mat( bed_sub = bed_subm, a = a, b = b, bs = bs, method = "peak_centered" )

hist_sanch_m <- non_scaled_mat( bed_sub = bed_subm, a = a, b = b, bs = bs, method = "single_anch" )
strandvec_sanch_m <- rep( bed_subm[[6]], times = ( a + b ) / bs )
      
mat_list <- bi_anch_mat( bed_sub = bed_subm, n = n )
hist_scaled_m <- mat_list[[1]]
ends_matm <- mat_list[[2]]
strandvec_scaled_m <- rep( bed_subm[[6]], times = n )
```
Okay, these look good, now I'm sure my reference points are correct.

```{r}
long_hist_center_p <- melt( hist_center_p, na.rm = TRUE )
```
I see, okay. It can only handle data.tables natively and matrices are depreciated.

So... does it have to be a formal matrix or could it be a data.table?

```{r}
long_hist_center_p <- melt(
  as.data.table( hist_center_p,
                 keep.rownames = TRUE 
                 ),
  na.rm = TRUE,
  id.vars = "rn"
  )

long_hist_sanch_p <- melt(
  as.data.table( hist_sanch_p,
                 keep.rownames = TRUE 
                 ),
  na.rm = TRUE,
  id.vars = "rn"
  )

long_hist_scaled_p <- melt(
  as.data.table( hist_scaled_p,
                 keep.rownames = TRUE 
                 ),
  na.rm = TRUE,
  id.vars = "rn"
  )

long_ends_mat_p <- melt(
  as.data.table( ends_matp,
                 keep.rownames = TRUE 
                 ),
  na.rm = TRUE,
  id.vars = "rn"
  )
###
long_hist_center_m <- melt(
  as.data.table( hist_center_m,
                 keep.rownames = TRUE 
                 ),
  na.rm = TRUE,
  id.vars = "rn"
  )

long_hist_sanch_m <- melt(
  as.data.table( hist_sanch_m,
                 keep.rownames = TRUE 
                 ),
  na.rm = TRUE,
  id.vars = "rn"
  )

long_hist_scaled_m <- melt(
  as.data.table( hist_scaled_m,
                 keep.rownames = TRUE 
                 ),
  na.rm = TRUE,
  id.vars = "rn"
  )

long_ends_mat_m <- melt(
  as.data.table( ends_matm,
                 keep.rownames = TRUE 
                 ),
  na.rm = TRUE,
  id.vars = "rn"
  )
```
Looks good, lets see if this causes a problem downstream.

```{r}
long_ends_center_p <- long_hist_center_p[[3]] + bs
long_ends_sanch_p <- long_hist_sanch_p[[3]] + bs
long_ends_center_m <- long_hist_center_m[[3]] + bs
long_ends_sanch_m <- long_hist_sanch_m[[3]] + bs
```
Let's get the bws
```{r}
bw_sub1 <- import(
      bw1, selection = GenomicSelection(
        "hg38", chrom = "chr19", colnames = "score") )
bw_sub2p <- import(
      bw2p, selection = GenomicSelection(
        "hg38", chrom = "chr19", colnames = "score") )
bw_sub2m <- import(
      bw2m, selection = GenomicSelection(
        "hg38", chrom = "chr19", colnames = "score") )
```

And bams
```{r}
sbp1 <- ScanBamParam( 
          which = GRanges(
            seqnames = "chr19",
            ranges = IRanges( start = 0, end = bv1$si@seqlengths[ bv1$si@seqnames == "chr19" ] ) ) )
sbp2 <- ScanBamParam( 
          which = GRanges(
            seqnames = "chr19",
            ranges = IRanges( start = 0, end = bv2$si@seqlengths[ bv2$si@seqnames == "chr19" ] ) ) )
```

```{r}
bam_aln1 <- GRanges( readGAlignments( bv1$bam_info, param = sbp1 ) )
bam_aln2 <- GRanges( readGAlignments( bv2$bam_info, param = sbp2 ) )
```
we need to revcomp the proseq bam. We'll skip sbp mode for now though.
```{r}
bam_aln2 <- bam_revcomp( bam_aln2, FALSE )
```

```{r}
gr_center_p <- GRanges( 
        seqnames = "chr19",
        ranges = IRanges( start = long_hist_center_p[[3]] + 1, end = long_ends_center_p ),
        strand = strandvec_sanch_p
      )
    gr_center_p$ID <-  long_hist_center_p[[1]]
    gr_center_p$BI <-  long_hist_center_p[[2]]

gr_sanch_p <- GRanges( 
        seqnames = "chr19",
        ranges = IRanges( start = long_hist_sanch_p[[3]] + 1, end = long_ends_sanch_p ),
        strand = strandvec_sanch_p
      )
    gr_sanch_p$ID <-  long_hist_sanch_p[[1]]
    gr_sanch_p$BI <-  long_hist_sanch_p[[2]]
    
gr_scaled_p <- GRanges( 
        seqnames = "chr19",
        ranges = IRanges( start = long_hist_scaled_p[[3]] + 1, end = long_ends_mat_p[[3]] ),
        strand = strandvec_scaled_p
      )
    gr_scaled_p$ID <-  long_hist_scaled_p[[1]]
    gr_scaled_p$BI <-  long_hist_scaled_p[[2]]
    
gr_center_m <- GRanges( 
        seqnames = "chr19",
        ranges = IRanges( start = long_hist_center_m[[3]] + 1, end = long_ends_center_m ),
        strand = strandvec_sanch_m
      )
    gr_center_m$ID <-  long_hist_center_m[[1]]
    gr_center_m$BI <-  long_hist_center_m[[2]]
    
gr_sanch_m <- GRanges( 
        seqnames = "chr19",
        ranges = IRanges( start = long_hist_sanch_m[[3]] + 1, end = long_ends_sanch_m ),
        strand =strandvec_sanch_m
      )
    gr_sanch_m$ID <-  long_hist_sanch_m[[1]]
    gr_sanch_m$BI <-  long_hist_sanch_m[[2]]
    
gr_scaled_m <- GRanges( 
        seqnames = "chr19",
        ranges = IRanges( start = long_hist_scaled_m[[3]] + 1, end = long_ends_mat_m[[3]] ),
        strand = strandvec_scaled_m
      )
    gr_scaled_m$ID <-  long_hist_scaled_m[[1]]
    gr_scaled_m$BI <-  long_hist_scaled_m[[2]]
```



```{r}
tmp_mat_center_p_bam1 <- bam_olaps( gr_center_p, bam_aln1, ignorestrand=TRUE, long_hist_center_p, debug=FALSE )
tmp_mat_center_m_bam1 <- bam_olaps( gr_center_m, bam_aln1, ignorestrand=TRUE, long_hist_center_m, debug=FALSE )

tmp_mat_sanch_p_bam1 <- bam_olaps( gr_sanch_p, bam_aln1, ignorestrand=TRUE, long_hist_sanch_p, debug=FALSE )
tmp_mat_sanch_m_bam1 <- bam_olaps( gr_sanch_m, bam_aln1, ignorestrand=TRUE, long_hist_sanch_m, debug=FALSE )

tmp_mat_scaled_p_bam1 <- bam_olaps( gr_scaled_p, bam_aln1, ignorestrand=TRUE, long_hist_scaled_p, debug=FALSE )
tmp_mat_scaled_m_bam1 <- bam_olaps( gr_scaled_m, bam_aln1, ignorestrand=TRUE, long_hist_scaled_m, debug=FALSE )
```

Okay, do these look correct in IGV?

Why do I have one less gene in minus? Was it just too short or something? Hmm.

Center p- 
ZNF526_42220311 2 reads at 90bp downstream of center (note, you will have different genes as I did not set a seed for this and I random sampled.) This is at position 42,220,502. Does this hold up in IGV? Yes it does.
Sanch p-
2 reads 50 bp upstream HNRNPUL1, at position 41,264,321. Yes, that also appears to match up correctly. (I probably shouldn't have used a heterochromatin marker anchored at genes to test this, though to be fair low signal is easier to count.)
Scaled p-
3 reads in bin 3 of WTIP. This is located at 34,481,954 and it looks like WTIP's scaled bins are around 100 bp. Accurate. I do see I didn't calc this as pairs, but I didn't tell it to using bam aln so that's not an error in the actual script.

center m- PRKD2, +10 bp, 1 read. Position 46,716,615. Accurate.
Sanch m- ZNF260, -50 bp, 2 reads. Position 36,528,301. Accurate.
Scaled m- C19orf12, bin 2, 2 reads. Position 29,715,039 with 17 bp bins. Accurate.

Awesome. Now let's check non-stranded bigwig.

```{r}
    olap <-  findOverlaps( gr_center_p, bw_sub1, ignore.strand = TRUE )
    bw_df <-  data.frame( bw_sub1[ subjectHits( olap ) ] )
    regions_df <-  data.frame( gr_center_p[ queryHits( olap ) ] )
    scored_regions <-  cbind( bw_df, regions_df )
    names( scored_regions ) <- c(
      "chrom", "bs", "be", "bin_w", "star", "score",
      names( scored_regions[ , 7:13 ] )
    )
    scored_regions$bs <- scored_regions$bs - 1 #0 index
    gs_before_bs <-  pmax( ( scored_regions$start - scored_regions$bs ), 0 )
    ge_after_be <-  pmax( ( scored_regions$be - scored_regions$end ), 0 )
    scored_regions$adjustor <- ( 
      scored_regions$bin_w - gs_before_bs - ge_after_be ) / scored_regions$bin_w
    scored_regions$adjusted_score <- scored_regions$score * scored_regions$adjustor
    setDT( scored_regions ) 
    tmp2 <- scored_regions[ ,
                            list( ( sum( adjusted_score ) / sum( adjustor ) ) ),
                            by = .( ID, BI )
                            ]
  hist <- dcast( tmp2, ID~BI, value.var = "V1" ) 
```


Okay good on the single now, go broad.

```{r}
tmp_mat_center_p_bw1 <- bw_olaps( gr_center_p, bw_sub1, ignorestrand=TRUE, debug=FALSE )
tmp_mat_center_m_bw1 <- bw_olaps( gr_center_m, bw_sub1, ignorestrand=TRUE, debug=FALSE )

tmp_mat_sanch_p_bw1 <- bw_olaps( gr_sanch_p, bw_sub1, ignorestrand=TRUE, debug=FALSE )
tmp_mat_sanch_m_bw1 <- bw_olaps( gr_sanch_m, bw_sub1, ignorestrand=TRUE, debug=FALSE )

tmp_mat_scaled_p_bw1 <- bw_olaps( gr_scaled_p, bw_sub1, ignorestrand=TRUE, debug=FALSE )
tmp_mat_scaled_m_bw1 <- bw_olaps( gr_scaled_m, bw_sub1, ignorestrand=TRUE, debug=FALSE )
```

Looks like while the BW is multiplied by a value, it is in general proportional to what we got from the bam, which is good and what was expected. There's a slightly more 'smooth' transition at the edges due to overlapping bins but this effect would probably disappear in locations with more reads.

Okay, what about a stranded bam? Does that work? 
```{r}
tmp_mat_center_p_bam2 <- bam_olaps( gr_center_p, bam_aln2, ignorestrand=FALSE, long_hist_center_p, debug=FALSE )
tmp_mat_center_m_bam2 <- bam_olaps( gr_center_m, bam_aln2, ignorestrand=FALSE, long_hist_center_m, debug=FALSE )

tmp_mat_sanch_p_bam2 <- bam_olaps( gr_sanch_p, bam_aln2, ignorestrand=FALSE, long_hist_sanch_p, debug=FALSE )
tmp_mat_sanch_m_bam2 <- bam_olaps( gr_sanch_m, bam_aln2, ignorestrand=FALSE, long_hist_sanch_m, debug=FALSE )

tmp_mat_scaled_p_bam2 <- bam_olaps( gr_scaled_p, bam_aln2, ignorestrand=FALSE, long_hist_scaled_p, debug=FALSE )
tmp_mat_scaled_m_bam2 <- bam_olaps( gr_scaled_m, bam_aln2, ignorestrand=FALSE, long_hist_scaled_m, debug=FALSE )
```

Okay why is only center p being weird? Well I restarted R and reran this and there was no problem, so I suspect I had a cache problem or was appending to an existing value somehow.
So.
Centerp - 7 reads at ZNF548 -10, which is located at 57,389,991. I might call that as 8 personally, but measuring IGV by eye is hard, so algorithm is probs correct. Strand is correct, but this isn't an area with signal on both strands. ANGPTL4 is though, and I confirmed it's correct.
We'll skip centerm and sanch p because they are a bit redundant with what we've already observed and just check sanch m.Looks correct at ERF!

What about scaled, and bigwigs?

scaled - accurate values, but I do need to keep the section I had at one point that flips scaled bams bins around left to right. Check to see if that's needed for BW too before doing it.

```{r}
tmp_mat_center_p_bw2 <- bw_olaps( gr_center_p, bw_sub2p, ignorestrand=FALSE, debug=FALSE )
tmp_mat_center_m_bw2 <- bw_olaps( gr_center_m, bw_sub2m, ignorestrand=FALSE, debug=FALSE )

tmp_mat_sanch_p_bw2 <- bw_olaps( gr_sanch_p, bw_sub2p, ignorestrand=FALSE, debug=FALSE )
tmp_mat_sanch_m_bw2 <- bw_olaps( gr_sanch_m, bw_sub2m, ignorestrand=FALSE, debug=FALSE )

tmp_mat_scaled_p_bw2 <- bw_olaps( gr_scaled_p, bw_sub2p, ignorestrand=FALSE, debug=FALSE )
tmp_mat_scaled_m_bw2 <- bw_olaps( gr_scaled_m, bw_sub2m, ignorestrand=FALSE, debug=FALSE )
```
Okay. Are they proportional to the bams? 

I 'think' the center ones look a bit different but that it's only attributable to the bw being single base pair instead of fullread, not an actual formal difference. Yeah, that's definitely it. Yeah, you see the same artifact in the Bws. You also see the strand flipping problem.

I do need to flip the minus bins.
